Syntax documentation + new /docs directory

formatc1702 opened the pull request:
Closes #107.

See current status in rendered form [here](https://github.com/formatc1702/WireViz/blob/feature/syntax-description/syntax.md).

formatc1702 commented:
Obviously this is a work in progress, so things are missing. Please add your comments keeping this in mind.

I thought about doing this via docstrings, but I think a separate documentation, for non-programmers, is a valuable addition.

kvid commented:
@slightlynybbled This could also be a place for the documentation you are missing, as I described in https://github.com/formatc1702/WireViz/pull/95#issuecomment-660966635. What do you think?

formatc1702 commented:
> @slightlynybbled This could also be a place for the documentation you are missing, as I described in [#95 (comment)](https://github.com/formatc1702/WireViz/pull/95#issuecomment-660966635). What do you think?

That's my intention with the `## Multiline strings` section 😉

slightlynybbled commented:
I see that you are generating a markdown-based documentation.  That is a great start and would certainly be a place to put the 'gotchas' that documentation typically contains.

What is 1yr vision for documentation?  `readme.md`?  Sphinx?  MkDocs?

formatc1702 commented:
> What is 1yr vision for documentation? `readme.md`? Sphinx? MkDocs?

I will have to think about it and read up on these systems before I choose. Perhaps we can open a new issue to discuss this?

aakatz3 commented:
What are your thoughts on including comprehensive documentation (and possibly snippets from the other markdown files) in the repo wiki, so it's easily browseable by first-time users? The markdown files look great, but it would be nice to either have a docs folder, a github-pages site, or a wiki with them for accessibility.

formatc1702 commented:
- Short term: Add `syntax.md` and link to it from the main `readme.md`
- Long term: Proper documentation as @slightlynybbled suggested, system is TBD
- Honestly, I wouldn't want to start a wiki; I'd rather have a dedicated (and mostly autogenerated?) GH Pages repo.

formatc1702 commented:
While there are a couple of open to-do's, the syntax description has reached a stage where I'm happy with the information that's included. It is rather dense, since it's not meant as a gentle intro, but as a reference guide.

I believe most of the WireViz syntax itself is described without any glaring omissions. Please provide feedback if
- a feature/attribute/behavior is missing
- something is described in an utterly incomprehensible, ambiguous or contradictory way
- you think it could be stylistically improved, language wise

Tyler-Ward commented:
The style and layout looks good to me, should work well as a reference for people writing input files.

There are a few references I spotted to features for 0.3 e.g #50 and #48, these are marked as todo in the code so I assume you are already aware of them.

formatc1702 commented:
> There are a few references I spotted to features for 0.3 e.g #50 and #48, these are marked as todo in the code so I assume you are already aware of them.

I've removed references to future features; it's probably best to add them to the documentation once they are actually implemented.

kvid commented on docs/syntax.md:
```diff
@@ -0,0 +1,340 @@
+  ![##000000](https://via.placeholder.com/15/000000/000000?text=+)
+  ...
+
+- `TEL` and `TELALT`  for [25-pair color code](https://en.wikipedia.org/wiki/Electronic_color_code#Color_band_system)
```
I assume the correct URL is `https://en.wikipedia.org/wiki/25-pair_color_code`, but double-check with the contributor. Where did you copy the existing link from?

formatc1702 replied:
You are correct. `README.md` links to the correct Wikipedia page.

kvid commented on docs/syntax.md:
````diff
@@ -0,0 +1,340 @@
+```yaml
+<str>   :  # unique connector designator/name
+  # general information about a connector (all optional)
+  type: <str>   
+  subtype: <str>   
````
- Same comment as above: Are there some limitations about a string like this?
- If a _designator_ can be any string, then these descriptions can be combined.
- If the only limitations are what can be specified in YAML, then a link to a description, might be enough.

formatc1702 commented on docs/buildscript.md:
```diff
@@ -0,0 +1,16 @@
+- Run `python build_examples.py compare` to compare generated files in all groups against the last commit.
+- Run `python build_examples.py clean` to delete generated files in all groups.
+- Run `python build_examples.py restore` to restore generated files in all groups from the last commit.
+- Edit 4: Append `-c` or `--compare-graphviz-output` to the `compare` command above to also compare the Graphviz output (default: False).
```
Will do. This commit is just a placeholder until I am inspired to write a more comprehensive doc :)

formatc1702 commented on docs/syntax.md:
````diff
@@ -0,0 +1,340 @@
+## Connector attributes
+
+```yaml
+<str>   :  # unique connector designator/name
````
This is probably a very deep rabbit hole, with all kinds of exceptions:
- numbers are accepted, but need to go in quotes so they are parsed as strings
- some special YAML characters (`-`, `:` and others) can be used by putting the string in quotes
- some characters will be accepted, but
  - might not show up in the graphical output
  - might mess with the BOM export
- some obscure cases might throw  errors

Honestly for now I would leave the explanation as-is, and let the user figure it out.
I do agree that it merits a more detailed description, but I would leave that for a later date.

kvid replied:
As this connector designator (and similar for cables) currently is used directly to name the Graphviz node, it must be case insensitive unique within the union of connectors and cables in the harness to avoid reference mix-ups similar to those discovered in #160.

formatc1702 replied:
I would like to leave it as-is now, hopefully fix #160 soon and then add "case-sensitive" to the syntax description

kvid commented on docs/CONTRIBUTING.md:
```diff
@@ -1,17 +1,23 @@
-1. Create new pull request
+- Fork this repository and clone it on your local machine.
+- Create a new feature branch on top of the `dev` branch.
+- Make any changes to the code on your feature branch.
```
Consider this line instead: `Commit your code changes to this feature branch.`

kvid commented on docs/CONTRIBUTING.md:
```diff
@@ -1,17 +1,23 @@
-1. Make any changes to the code on the `dev` branch
-1. Push your changes to your fork
-1. Create new pull request
+- Fork this repository and clone it on your local machine.
+- Create a new feature branch on top of the `dev` branch.
+- Make any changes to the code on your feature branch.
+- Push the changes to your fork.
+- Create new pull request.
```
Why removing the order numbers? The order is important.

formatc1702 commented:
Added your suggestions!

formatc1702 commented on docs/syntax.md:
```diff
@@ -307,6 +309,27 @@ Supported color codes:
+  caption: <str>     # text to display below the image
+  width: <int>       # range: 1~65535; unit: ???
+  height: <int>      # range: 1~65535; unit: ???
+  fixedsize: <bool>  # ???
```
I'm a bit confused about this attribute... even after reading the GraphViz docs

formatc1702 commented on docs/syntax.md:
```diff
@@ -307,6 +309,27 @@ Supported color codes:
+  src: <path>        # path to the image file
+  # optional parameters:
+  caption: <str>     # text to display below the image
+  width: <int>       # range: 1~65535; unit: ???
+  height: <int>      # range: 1~65535; unit: ???
```
In xxx you mentioned the unit to be `pt`... I'm not too sure. The [GraphViz docs](https://graphviz.org/doc/info/attrs.html#d:image) mention `pt` as the default for SVG files; for bitmaps, they only mention

> The file must contain the image size information. This is usually trivially true for the bitmap formats.

And I wonder if it is then in `px`?

kvid replied:
How about the simplified version here:
```
image:
  src: <path>        # path to the image file
  # optional parameters:
  caption: <str>     # text to display below the image
  width: <int>       # scale image to width in points
  height: <int>      # scale image to height in points
  # image might get distorted if both width and height are set
```
[More advanced scaling](advanced.md) is also possible, but seldom needed.

I suggest adding the advanced description below to a new `.md` file:

## Advanced Image Scaling

In rare cases when the ordinary image scaling functionality is insufficient, a couple of extra optional image parameters offer extra scaling functionality when combined with the image dimension parameters `width` and `height`:
- `scale: <str>` is default `false` if no dimension is specified, or `true` if only one dimension is specified, or `both` if both dimensions are specified.
- `fixedsize: <bool>` is default `false` when no dimension is specified or if `scale` is specified, and `true` otherwise.
- If `fixedsize` is true and only one dimension is specified, then the other dimension is calculated using the image aspect ratio. If reading the aspect ratio fails, then 1:1 ratio is assumed.

### The effect of `fixedsize` boolean values

- When `false`, any `width` or `height` values are _minimum_ values used to expand the image cell size for more available space, but cell contents or other size demands in the table might expand this cell even more than specified by `width` or `height`.
- When `true`, both `width` and `height` specifiers are required and specify the fixed size of the image cell, distorting any image inside if it don't fit. Any borders are normally drawn around the fixed size, and therefore, WireViz enclose the image cell in an extra table without borders when `fixedsize` is true to keep the borders around the outer non-fixed cell.

### The effect of `scale` string values:

- When `false`, the image is not scaled.
- When `true`, the image is scaled proportionally to fit within the available image cell space.
- When `width`, the image width is expanded (height is normally unchanged) to fill the available image cell space width.
- When `height`, the image height is expanded (width is normally unchanged) to fill the available image cell space height.
- When `both`, both image width and height are expanded independently to fill the available image cell space.

In all cases (except `true`) the image might be distorted when a specified fixed image cell size limits the available space to less than what an unscaled image needs.

In the WireViz diagrams there are no other space demanding cells in the same row, and hence, there are never extra available image cell space height unless a greater image cell `height` also is specified.

### How Graphviz support this image scaling

The connector and cable nodes are rendered using a HTML `<table>` containing an image cell `<td>` with `width`, `height`, and `fixedsize` attributes containing an image `<img>` with `src` and `scale` attributes. See also the [Graphviz doc](https://graphviz.org/doc/info/shapes.html#html), but note that WireViz uses default values as described above.

formatc1702 commented on docs/syntax.md:
```diff
@@ -307,6 +309,27 @@ Supported color codes:
+  width: <int>       # range: 1~65535; unit: ???
+  height: <int>      # range: 1~65535; unit: ???
+  fixedsize: <bool>  # ???
+  scale: <str>       # scaling behavior; possible values:
+                     # false   does not scale the image
+                     # true    scale image proportionally
+                     #         to fit within given max. width and height
+                     # width   scale image proportionally to given width
+                     # height  scale image proportionally to given height
+                     # both    resize to given width and height
```
Please let me know if I summarized the options correctly

formatc1702 commented:
> Have you considered using normal markdown instead of the limited fixed character width formatting to improve readability and enable internal anchor links?

Yes. I quite like the super compact format I use now, but for this kind of detailed explanation of the behavior of every possible parameter value, it's probably better to have it as regular text.

kvid commented:
> > Have you considered using normal markdown instead of the limited fixed character width formatting to improve readability and enable internal anchor links?
> 
> Yes. I quite like the super compact format I use now, but for this kind of detailed explanation of the behavior of every possible parameter value, it's probably better to have it as regular text.

I guess we should start with a simplified description (as in https://github.com/formatc1702/WireViz/pull/153#issuecomment-674536040) of the alternatives when not specifying `scale` or `fixedsize`, and that would be enough for most users. I suggest moving my advanced summary suggestions above into a separate .md file for advance usage where all attributes are described together with their default values as a reference for the few users that might need it.

If you agree to my last commits, then this can be a description of default values of image attributes:

Unless they are specified in the YAML file, these attributes get their default values as described below:
- `scale` = "false" if no dimension is specified, or "true" if only one dimension is specified, or "both" if both dimensions are specified.
- `fixedsize` = True if `scale` is "true" or "both", or False otherwise.
- If `fixedsize` is True and only one dimension is specified, then the other dimension is calculated using the image aspect ratio. If reading the aspect ratio fails, then 1:1 ratio is assumed.

kvid commented on docs/buildscript.md:
```diff
@@ -0,0 +1,16 @@
+- Run `python build_examples.py clean` to delete generated files in all groups.
+- Run `python build_examples.py restore` to restore generated files in all groups from the last commit.
+- Append `-c` or `--compare-graphviz-output` to the `compare` command above to also compare the Graphviz output (default: False).
+- Append `-g` or `--groups` followed by space separated group names to any command above, and the set of generated files affected by the command will be limited to the selected groups.
```
I suggest ending this paragraph with the possible group names:

- Append `-g` or `--groups` followed by space separated group names to any command above, and the set of generated files affected by the command will be limited to the selected groups (names: `examples tutorial demos`).

formatc1702 replied:
Done.

kvid commented on docs/syntax.md:
```diff
@@ -0,0 +1,363 @@
+
+## Inheritance
+
+Add link to YAML spec.
```
Remember adding a link here.

formatc1702 replied:
I'll add a link to [this page](https://blog.daemonl.com/2016/02/yaml.html) which explains YAML anchors very concisely.
Since it's supposed to be a WireViz syntax description, and not a YAML tutorial, I'll keep it short and sweet for now, with a reference to demo02 for an example. The `--prepend-file` function is not really part of the syntax, so I won't include it here.

My hope is that in the future, someone with more writing skills than me can contribute a full, from-start-to-finish WireViz tutorial for first-time users with no previous experience with YAML, explaining all these features in a friendly and understandable way. But that is an issue for another day ;)

kvid reviewed (changes requested):
Some of my older suggestions are obsolete if accepting these new suggestions.

kvid commented on docs/README.md:
```diff
@@ -26,23 +32,12 @@ WireViz is a tool for easily documenting cables, wiring harnesses and connector
 
```
The standard documents are not available for free, but they have a preview showing the first pages, including their scope.

[IEC 60757:1983](https://webstore.iec.ch/publication/3406) preview:
- Title: Code for designation of colours
- Scope: This standard applies to the text of descriptions, drawings, markings, etc., in the
electrotechnical field.

[IEC 60062:2016+AMD1:2019 CSV Consolidated version](https://webstore.iec.ch/publication/65655) preview:
- Title: Marking codes for resistors and capacitors
- Scope: This International Standard specifies designation and marking codes for capacitors and resistors. It provides coding methods for the resistance or capacitance value and its tolerance, including colour coding for resistors. It provides coding for parameters specific either to capacitors, like e.g. the dielectric material, or to resistors, like e.g. the temperature coefficient of resistance (TCR). It also provides date code systems suitable for the marking of small components.

formatc1702 replied:
I guess IEC 60757 is then probably the right choice here!

kvid commented on docs/syntax.md:
```diff
@@ -0,0 +1,363 @@
+  ![##8000ff](https://via.placeholder.com/15/8000ff/000000?text=+)
+  ...
+
+- `IEC` for [IEC 62](https://en.wikipedia.org/wiki/Electronic_color_code#Color_band_system) ("ROY G BIV")
```
See earlier comments today about the IEC standards: https://github.com/formatc1702/WireViz/pull/111#discussion_r477001676 and https://github.com/formatc1702/WireViz/pull/111#discussion_r477495307

kvid commented on docs/README.md:
```diff
@@ -1,9 +1,15 @@
 # WireViz
 
+
+[![PyPI - Version](https://img.shields.io/pypi/v/wireviz.svg?colorB=blue)](https://pypi.org/project/wireviz/)
+[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/wireviz.svg?)](https://pypi.org/project/wireviz/)
+[![PyPI - Downloads](https://img.shields.io/pypi/dm/wireviz)](https://pypi.org/project/wireviz/)
```
Do you want any other badges from https://shields.io/
- https://img.shields.io/github/issues/formatc1702/WireViz
- https://img.shields.io/github/forks/formatc1702/WireViz
- https://img.shields.io/github/stars/formatc1702/WireViz
- https://img.shields.io/github/license/formatc1702/WireViz
- https://img.shields.io/twitter/url?url=https%3A%2F%2Fgithub.com%2Fformatc1702%2FWireViz%2F

formatc1702 replied:
Thanks. Since issues, forks, stars and license are already clearly displayed on the project's homepage, I don't think this adds too much valuable information.

kvid commented on docs/syntax.md:
```diff
@@ -0,0 +1,363 @@
+
+## BOM items
+
+Connectors (both regular, and auto-generated), cables, and wires of a bundle are automatically added to the BOM,
+unless the `ignore_in_bom` attribute is set to `true`.
```
`ignore_in_bom` is not yet implemented. See #50 and #115.

Tyler-Ward replied:
If the ignore_in_bom feature is wanted for 0.2 it can be fairly easily merged in on its own as it is the first commit in the branch for #115 (716cc06ea9be0df11b61b8f1768e71625508a46d)

kvid reviewed (changes requested):
Suggesting additions to Contribution Guidelines

formatc1702 commented on docs/README.md:
```diff
@@ -26,23 +32,12 @@ WireViz is a tool for easily documenting cables, wiring harnesses and connector
 
```
I'll keep it as is... nobody has complained, and I'd like to address this in the near-ish future :)

formatc1702 replied:
Update: I've decided to comment out the line, with a reference to #158 so as not to forget to re-include it in the future.

formatc1702 commented on docs/README.md:
```diff
@@ -26,23 +32,12 @@ WireViz is a tool for easily documenting cables, wiring harnesses and connector
 
```
It was meant to mean it was not only for simple cables where all wires go from one connector A on one side to another connector B on the other, but allows for more complex harnesses... I agree it's not very clear, will think about a change.

formatc1702 commented on docs/syntax.md:
```diff
@@ -0,0 +1,363 @@
+
+## Multiline strings
+
+Connectors accept multiline strings in the `type`, `subtype` and `notes` attributes.
+Cables accept multiline strings in the `type` and `notes` attributes.
```
> > Alternatively, we could consider using something like `<mlstr>` as type where a multiline string is accepted.
> 
> After adding semanticly named type aliases in PR #163, I now suggest we do something similar here. If PR #163 is accepted, I suggest we use the same type names here as in `DataClasses.py`. See also [#163 (comment)](https://github.com/formatc1702/WireViz/pull/163#discussion_r487463273) about how PR #164 might affect this.

Let's think about this in the future, once those PRs are actually merged :) For now, a list of multiline-supporting attributes is enough.

formatc1702 commented:
I've included a lot of your suggestions, and commented on the ones where I think it's worth leaving as-is for now and expanding in the future, so as not to overthink it in this still early stage :) I hope you agree on these points, looking forward to finishing up and merging the current version of the docs.

formatc1702 commented:
If there are no major disagreements, @kvid please give green light to squash and merge :)

Otherwise let me know, I've hopefully addressed all your comments. Thanks!

kvid commented:
> If there are no major disagreements, @kvid please give green light to squash and merge :)
> 
> Otherwise let me know, I've hopefully addressed all your comments. Thanks!

I'm sorry, but I don't have a PC where I am right now. I only have a phone with a small screen. I have to wait until later today before I'm able to read your latest effort properly. However, I do know about a few issues:
- Have you updated the doc with the latest command line option changes of both wireviz and build_examples? Run both of them with `--help` to see the new options.
- I have made a list at home of PRs that are missing from the [CHANGELOG.md](https://github.com/formatc1702/WireViz/blob/feature/syntax-description/docs/CHANGELOG.md), and at least one bug in the existing references.

kvid commented:
> * I have made a list at home of PRs that are missing from the [CHANGELOG.md](https://github.com/formatc1702/WireViz/blob/feature/syntax-description/docs/CHANGELOG.md), and at least one bug in the existing references.

Please browse quickly through the references to check that I didn't misinterpreted any of them.

Missing change log entries (feel free to join minor or related changes together):
- New colors #103
- Replace unicode character with superscript `<sup>2</sup>` #95 
- Bugfix #69
- Change code for silver and gold #113
- Short BOM names #114 #121
- Bugfix #125
- New build_examples features #118
- Padding of single-color wires globally #131 #132 
- Bugfix #144 #145
- Bugfix #139 #140 
- New cable.color + HTML whitespace #66 #136 
- GitHub Linguist #154
- Command line options #167 #173 
- Images #27 #153 

Bad existing reference:
- Handle line breaks #49 (NOT #63)

TODO:
- Documentation #107 #111

kvid commented:
> If there are no major disagreements, @kvid please give green light to squash and merge :)
> 
> Otherwise let me know, I've hopefully addressed all your comments. Thanks!

I can't find any reaction (comment or code change) related to my https://github.com/formatc1702/WireViz/pull/111#discussion_r487471800 - did you consider this one?

formatc1702 commented:
Wow, that's an exhaustive list, amazing! When I started the changelog, I thought I'd just include the most relevant changes, but perhaps it's good practice to include the complete list. Thanks. I will do this tomorrow.

> I can't find any reaction (comment or code change) related to my https://github.com/formatc1702/WireViz/pull/111#discussion_r487471800 - did you consider this one?

I've updated `buildscript.md` to reflect the changes suggested in your linked comment, and updated `README.md` to show `wireviz.py`'s CLI options.

formatc1702 commented on docs/README.md:
````diff
@@ -106,29 +134,25 @@ mywire.bom.tsv    BOM (bill of materials) as tab-separated text file
 mywire.html       HTML page with wiring diagram and BOM embedded
 ```
````
Done.

kvid reviewed (changes requested):
Be aware that https://github.com/formatc1702/WireViz/pull/111#issuecomment-708602324 has been edited since your last message in this PR.

kvid reviewed (changes requested):
I overlooked this one yesterday.

formatc1702 commented:
I've added all the issue/PR links you provided. Hopefully we can close this now, ideally I'd like to package up v0.2 today since I have some time.

kvid commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,43 @@
+  - You might then find an existing answer or suggested solution to your issue, possibly also an existing PR you can test.
+  - When finding existing issues that seem related to your issue, please include references (# followed by issue number) to related issues in your new issue description, or if a very similar issue is still open, consider adding a comment in that issue instead of creating a new one.
+- When appropriate, please prefix your issue title with one of these category prefixes followed by a space:
+  - **[bug]** When the issue seems to be caused by a bug.
+  - **[feature]** When requesting a feature change or new feature.
+  - **[internal]** When suggesting code improvements that doesn't change any output.
+  - **[doc]** For documentation issues.
+  - **[meta]** For issues about the development or contribution process.
```
We have lately also used some other categories, not listed here, but I assume this doesn't have to be a complete list... no changes required.

formatc1702 replied:
Yep. This list should cover most cases. Anyway it's not a super strict rule, but it helps reduce the chaos a bit.

kvid commented on docs/CHANGELOG.md:
```diff
@@ -0,0 +1,45 @@
+- Add support for multicolored wires ([#12](https://github.com/formatc1702/WireViz/issues/12), [#17](https://github.com/formatc1702/WireViz/pull/17), [#96](https://github.com/formatc1702/WireViz/pull/96), [#131](https://github.com/formatc1702/WireViz/issues/131), [#132](https://github.com/formatc1702/WireViz/pull/132))
+- Add support for images ([#27](https://github.com/formatc1702/WireViz/issues/27), [#153](https://github.com/formatc1702/WireViz/pull/153))
+- Add ability to export data directly to other programs ([#55](https://github.com/formatc1702/WireViz/pull/55))
+- Add support for line breaks in various fields ([#49](https://github.com/formatc1702/WireViz/issues/49))
```
#64 solves #49

kvid reviewed (approved):
Thank you for accepting almost all of my suggestions. And thank you for the effort creating this nice documentation. Too many projects neglect documentation in the early stages of a project.

formatc1702 commented:
> Thank you for accepting almost all of my suggestions. And thank you for the effort creating this nice documentation. Too many projects neglect documentation in the early stages of a project.

1) It's really annoying having to hand-pick and document every little change and keep things tidy...
2) But imagine having to do that 1, 2, 6 or 12 months from now! 😅 
3) Thanks for taking care of 99% of the work here.

Please have a look at #178 in case I'm forgetting something... I'm trying to make a foolproof checklist, which I can then reuse on future releases, so feel free to add your thoughts.

I'm excited to finish this PR but will wait until your final 👍 .

kvid commented:
> > Thank you for accepting almost all of my suggestions. And thank you for the effort creating this nice documentation. Too many projects neglect documentation in the early stages of a project.
> 
> 1. It's really annoying having to hand-pick and document every little change and keep things tidy...
> 2. But imagine having to do that 1, 2, 6 or 12 months from now! 😅

I suggest adding a minor entry to the change log immediately after merging any PR. It should be much easier when it is fresh in mind, instead of searching all PRs just before a release. 😃 

BTW, you just merged my PR #177 in case that fits into one of the change log items - perhaps the item about "Improve HTML output".

> 3. Thanks for taking care of 99% of the work here.

Not at all. A lot of minor suggestions only. You have put everything together.

> 
> Please have a look at #178 in case I'm forgetting something... I'm trying to make a foolproof checklist, which I can then reuse on future releases, so feel free to add your thoughts.
> 
> I'm excited to finish this PR but will wait until your final 👍 .

formatc1702 commented:
> I suggest adding a minor entry to the change log immediately after merging any PR. It should be much easier when it is fresh in mind, instead of searching all PRs just before a release. 😃
> 
> BTW, you just merged my PR #177 in case that fits into one of the change log items - perhaps the item about "Improve HTML output".

🤦  Yes, this should become standard practice. Time for a quick force-push to `dev`... 🤫 You didn't see anything!
